pkg/envoy/sds/builder.go

```diff
@@ -0,0 +1,172 @@
+	"github.com/openservicemesh/osm/pkg/service"
+)
+
+type sdsBuilder struct {
+	proxy *envoy.Proxy
+
+	// Service certificate for this proxy
+	proxyCert *certificate.Certificate
+
+	// Requested certs to generate secrets for
+	requestedCerts []string
+
+	trustDomain string
+
+	ListServiceIdentitiesForService func(service.MeshService) []identity.ServiceIdentity
+}
+
+func newSdsBuilder() *sdsBuilder {
+	return &sdsBuilder{}
+}
+
+func (b *sdsBuilder) SetProxy(proxy *envoy.Proxy) {
+	b.proxy = proxy
+}
+
+func (b *sdsBuilder) SetProxyCert(cert *certificate.Certificate) {
+	b.proxyCert = cert
+}
+
+func (b *sdsBuilder) SetRequestedCerts(requestedCerts []string) {
+	b.requestedCerts = requestedCerts
+}
+
+func (b *sdsBuilder) SetTrustDomain(trustDomain string) {
+	b.trustDomain = trustDomain
+}
+
+func (b *sdsBuilder) SetListServiceIdentitiesForService(listServiceIdentitiesForService func(service.MeshService) []identity.ServiceIdentity) {
+	b.ListServiceIdentitiesForService = listServiceIdentitiesForService
+}
+
+// Build generates SDS Secret Resources based on requested certs in the DiscoveryRequest
+func (b *sdsBuilder) Build() ([]types.Resource, error) {
+	var sdsResources []types.Resource
+
+	for _, envoyProto := range b.getSDSSecrets(b.proxyCert, b.requestedCerts, b.proxy) {
+		sdsResources = append(sdsResources, envoyProto)
+	}
+	return sdsResources, nil
+}
+
+func (b *sdsBuilder) getSDSSecrets(cert *certificate.Certificate, requestedCerts []string, proxy *envoy.Proxy) (certs []*xds_auth.Secret) {
+	// requestedCerts is expected to be a list of either of the following:
+	// - "service-cert:namespace/service-account"
+	// - "root-cert-for-mtls-outbound:namespace/service"
+	// - "root-cert-for-mtls-inbound:namespace/service-account"
+
+	// The Envoy makes a request for a list of resources (aka certificates), which we will send as a response to the SDS request.
+	for _, requestedCertificate := range requestedCerts {
+		sdsCert, err := secrets.UnmarshalSDSCert(requestedCertificate)
+		if err != nil {
+			log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrUnmarshallingSDSCert)).
+				Msgf("Invalid resource kind requested: %q", requestedCertificate)
+			continue
+		}
+
+		log.Debug().Str("proxy", proxy.String()).Msgf("Proxy requested cert %s", requestedCertificate)
+
+		switch sdsCert.CertType {
+		// A service certificate is requested
+		case secrets.ServiceCertType:
+			envoySecret, err := b.getServiceCertSecret(cert, requestedCertificate)
+			if err != nil {
+				log.Error().Err(err).Str(errcode.Kind, errcode.GetErrCodeWithMetric(errcode.ErrGettingServiceCertSecret)).
+					Str("proxy", proxy.String()).Msgf("Error getting service cert %s for proxy", requestedCertificate)
+				continue
+			}
+			certs = append(certs, envoySecret)
+
+		// A root certificate used to validate a service certificate is requested
+		case secrets.RootCertTypeForMTLSInbound, secrets.RootCertTypeForMTLSOutbound:
```

Comment:
as a side note: we can notice that every single rootcert is identical.

These get picked based on the name that is set in `envoy.GetUpstreamTLSContext` in each cluster config.

With this refactoring we could also switch these names to be a static name: `osm-root-cert`, and have just 1 single root cert for validation, instead of one identical root cert for each individual service.

Not necessary in this PR, but worth exploring to understand ramifications :)